### ISSUE TITLE
[Feature] onSuiteStart/onSuiteEnd hooks for Reporter

### ISSUE THREAD

**Issue by alex-popov-tech** [microsoft/playwright#23865]
Hello!

I'm trying to build custom reporter, and I found myself missing two hooks - `onSuiteStart` and `onSuiteEnd`, especially last one.

It is pretty hard and resource-consuming (and maybe just won't work with parallelisation ) to try to mimic these hooks in `onTestEnd`.

**Comment by dgozman** [comment]
@alex-popov-tech As you've noticed, `onSuiteStart` and `onSuiteEnd` is not really compatible with parallelization. You can implement the logic you'd like with `onTestEnd`, for example:

```ts
onTestEnd(test) {
  const suite = test.parent;
  const tests = suite.allTests();
  const allTestsFinished = tests.every(t => t.results.length > 0);
  if (allTestsFinished) {
    // handle suite finish
  }
}
```

Let me know whether this helps.

**Comment by alex-popov-tech** [comment]
@dgozman thanks, that is a bit better.
but question is still - is `allTests()` multithread safe? f.e. - can there be situation when two parallel tests finishing nearly at the same time, and in `onTestEnd` of both of them the other one is considered not finished yet?

in other words - if i would use parallelisation, using `onTestEnd` will gurantee me that i will catch all suite ends through `test.parent.allTests()`?

**Comment by dgozman** [comment]
> in other words - if i would use parallelisation, using onTestEnd will gurantee me that i will catch all suite ends through test.parent.allTests()?

Yes, JavaScript is single-threaded, so you will always have consistent data in all your reporter methods.

**Comment by alex-popov-tech** [comment]
@dgozman but you parallelise using workers, which have their own contexts and event loops, don't they?

**Comment by dgozman** [comment]
@alex-popov-tech Reporter runs in the main runner process and all events from multiple parallel workers arrive to that single place. You don't have to worry about these details though, API is consistent so you can naturally use it.

**Comment by dgozman** [comment]
@alex-popov-tech That's because `onSuiteEnd` is not really helpful in general case. For example, with unlucky test scheduling/retries, all suites may run 99% of their tests first, and then all finish at the very end with their last test running. You can be sure that all suites have finished in `onEnd()` 😄  With nested suites, calling `onSuiteEnd` could get very noisy.

If you have a usecase that would benefit many users, we might consider adding this API, but so far there have been none.

**Comment by alex-popov-tech** [comment]
@dgozman usecase is super simple, it is about `Test` and `Suite` having same consise lifecycle.
Like, in html reporter, i might want to track this progress, to show users that:

1. `Test` is scheduled, but not started,
2. `Test` is running
3. `Test` is resolved with status

with pretty loading wheels and so on. but since tests are unseparateble from suits, it makes little sense to add `Test` status tracking without same thing for `Suite`, because to see `Test` users need to open its `Suite`

**Comment by dgozman** [comment]
@alex-popov-tech Sounds good. Let me leave this issue open to gauge interest.

**Comment by pavelfeldman** [comment]
#### Why was this issue closed?

Thank you for your involvement. This issue was closed due to limited engagement (upvotes/activity), lack of recent activity, and insufficient actionability. To maintain a manageable database, we prioritize issues based on these factors.

If you have additional information not present in this issue that you think will help prioritizing it, please open a new issue and reference this one. More support or clarity on its necessity may prompt a review. Your understanding and cooperation are appreciated.
